File: python/flintdb/backup.py
class Backup:
    """Utility class for handling backup and recovery of database."""
    @staticmethod
    def dump(database: "Database", file: str) -> None:
        """Backup database to a zip file.

        Args:
            database: The instance of Database to backup.
            file: The filesystem path to the zip file.
        """
        from zipfile import ZipFile, ZIP_DEFLATED
        with ZipFile(file, mode="w", compression=ZIP_DEFLATED, compresslevel=9) as zip:
            zip.mkdir(database.name())
            zip.write(database.folder() / ".metadata", arcname=database.name() + "/.metadata")
            for table in database.tables():
                zip.mkdir(database.name() + "/" + table.name())
                zip.write(table.folder() / ".metadata", arcname=database.name() + "/" + table.name() + "/.metadata")
                for row in table.rows():
                    zip.write(row.file(), arcname=database.name() + "/" + table.name() + "/" + row.id() + ".ndjson")

    @staticmethod
    def load(file: str, storage: str) -> None:
        """Recover database backup from zip file.

        Args:
            file: The filesystem path to the zip file.
            storage: The directory path where the database folder should reside.
        """
        from zipfile import ZipFile
        with ZipFile(file, mode="r") as zip:
            zip.extractall(storage)

File: php/flintdb/Backup.php
<?php

namespace FlintDB;

use RuntimeException,
    ZipArchive;

class Backup {
  /**
   * Backup database to zip file
   *
   * @param Database $database
   * @param string $file
   * @return void
   * @throws RuntimeException
   */
  public function dump( Database $database, string $file ): void {
    $zip = new ZipArchive;
    if ( $zip->open( $file, ZipArchive::CREATE | ZipArchive::OVERWRITE ) !== true ) {
      throw new RuntimeException( 'Failed to open the ZIP file' );
    }
    
    $dir = $database->storage();
    $zip->addEmptyDir( $database->name() );
    $zip->addFile( $dir . '/' . $database->name() . '/.metadata', $database->name() . '/.metadata' );
    foreach ( $this->database->yield_tables() as $table ) {
      $folder = $database->name() . '/' . $table->name();
      $zip->addEmptyDir( $folder );
      $zip->addFile( $dir . '/' . $folder . '/.metadata', $folder . '/.metadata' );
      foreach ( $table->yield_rows() as $row ) {
        $file = $folder . '/' . $row->id() . '.ndjson';
        $zip->addFile( $dir . $file, $file );
      }
    }
    
    $zip->close();
  }
  
  /**
   * Recover database from zip file
   *
   * @param string $file
   * @param string $storage
   * @return void
   * @throws RuntimeException
   */
  public static function load( string $file, string $storage ): void {
    $zip = new ZipArchive;
    if ( $zip->open( $file ) !== true ) {
      throw new RuntimeException( 'Failed to open the ZIP file' );
    }
    
    $zip->extractTo( $storage );
    $zip->close();
  }
}
